libalpm/scripts/mkinitcpio-install

```diff
@@ -0,0 +1,19 @@
+#!/bin/bash
+
+args=('-p')
+
+while read -r line; do
+    if [[ $line = usr/lib/modules/+([^/])/pkgbase ]]; then
+        pkgbase=$(< "/$line")
+        if [[ ! -e /etc/mkinitcpio.d/"${pkgbase}".preset ]]; then
+            sed "s|%PKGBASE%|${pkgbase}|g" /usr/share/mkinitcpio/hook.preset | install -Dm644 /dev/stdin \
+              /etc/mkinitcpio.d/"${pkgbase}".preset
+        fi
+        install -Dm644 $(dirname $line)/vmlinuz /boot/vmlinuz-"${pkgbase}"
+    else
+        args=('-P') # all presets
+        break
+    fi
+done
+
+mkinitcpio "${args[@]}" "${pkgbase}"
```

Comment:
This will run mkinitcpio for at most one `pkgbase`. I think what you want to start with an empty `args=()` and add `args+=(-p "$pkgbase")` each loop.

Also, I think the `if` check and the `else` branch doing `-P` is pretty pointless as you will always have a list `pkgbase` files as input.

Reply:
```bash
#!/bin/bash -e

args=()

while read -r line; do
    pkgbase=$(<"$line")
    preset="/etc/mkinitcpio.d/${pkgbase}.preset"
    if [[ ! -e $preset ]]; then
        sed "s|%PKGBASE%|${pkgbase}|g" /usr/share/mkinitcpio/hook.preset \
            | install -Dm644 /dev/stdin "$preset"
    fi
    install -Dm644 "${line%/pkgbase}/vmlinuz" "/boot/vmlinuz-${pkgbase}"
done

mkinitcpio "${args[@]}"
```

Reply:
See the hook, it triggers on usr/lib/modules/*/pkgbase as well as on Package = mkinitcpio*. So, -P is there to trigger when mkinitcpio itself is being upgraded. Using mkinitcpio* also has the benefit of regenerating all presets when mkinitcpio-busybox or mkinitcpio-nfs-utils are upgraded as well. Not to mention my own packages that are on community.

Reply:
Ah, right, the hook also matches on `mkinitcpio-*` packages... Still, we need to make sure to process all kernel targets.

In addition, I think it looks more correct for the hook to match on `vmlinuz` instead of `pkgbase`.

```bash
#!/bin/bash -e

args=()
all=0

while read -r line; do
    if [[ $line != */vmlinuz ]]; then
        all=1
        continue
    fi

    pkgbase=$(<"${line%/vmlinuz}/pkgbase")
    [[ -n ${pkgbase} ]]

    preset="/etc/mkinitcpio.d/${pkgbase}.preset"
    if [[ ! -e $preset ]]; then
        sed "s|%PKGBASE%|${pkgbase}|g" /usr/share/mkinitcpio/hook.preset \
            | install -Dm644 /dev/stdin "$preset"
    fi

    install -Dm644 "${line}" "/boot/vmlinuz-${pkgbase}"
    args+=(-p "${pkgbase}")
done

if (( all )); then
    args=(-P)
fi

mkinitcpio "${args[@]}"
```

```ini
[Trigger]
Type = Package
Operation = Install
Operation = Upgrade
Target = mkinitcpio*

[Trigger]
Type = File
Operation = Install
Operation = Upgrade
Target = usr/lib/modules/*/vmlinuz

[Action]
Description = Updating boot kernels and initcpios...
When = PostTransaction
Exec = /usr/share/libalpm/scripts/mkinitcpio-install
NeedsTargets
```

Reply:
I've also just tested this hook installing two kernels at the same time, and it's working just for the first of them. The mkinitcpio run itself must be inside the loop.

Reply:
Implemented by dfa609a6ee5f75af1aa5f94d331aea60b6d076ca